src/gas/GasFee.controller.ts

```diff
@@ -0,0 +1,212 @@
+  suggestedMaxPriorityFeePerGas: string; // a GWEI hex number
+  suggestedMaxFeePerGas: string; // a GWEI hex number
+}
+
+/**
+ * @type GasFeeEstimates
+ *
+ * Data necessary to provide multiple GasFee estimates, and supporting information, to the user
+ *
+ * @property low - A GasFee for a minimum necessary combination of tip and maxFee
+ * @property medium - A GasFee for a recommended combination of tip and maxFee
+ * @property high - A GasFee for a high combination of tip and maxFee
+ * @property estimatedNextBlockBaseFee - An estimate of what the base fee will be for the pending/next block. A GWEI hex number
+ */
+
+export interface GasFeeEstimates {
+  low: Eip1559GasFee;
+  medium: Eip1559GasFee;
+  high: Eip1559GasFee;
+  estimatedBaseFee: string;
+}
+
+const metadata = {
+  legacyGasPriceEstimates: { persist: true, anonymous: false },
+  gasFeeEstimates: { persist: true, anonymous: false },
+};
+
+/**
+ * @type GasFeeState
+ *
+ * Gas Fee controller state
+ *
+ * @property legacyGasPriceEstimates - Gas fee estimate data using the legacy `gasPrice` property
+ * @property gasFeeEstimates - Gas fee estimate data based on new EIP-1559 properties
+ */
+export type GasFeeState = {
+  gasFeeEstimates: GasFeeEstimates | Record<string, never>;
+};
+
+const name = 'GasFeeController';
+
+export type GasFeeStateChange = {
+  type: `${typeof name}:stateChange`;
+  payload: [GasFeeState, Patch[]];
+};
+
+export type GetGasFeeState = {
+  type: `${typeof name}:getState`;
+  handler: () => GasFeeState;
+};
+
+const defaultState = {
+  gasFeeEstimates: {},
+};
+
+/**
+ * Controller that retrieves gas fee estimate data and polls for updated data on a set interval
+ */
+export class GasFeeController extends BaseController<typeof name, GasFeeState> {
+  private intervalId?: NodeJS.Timeout;
+
+  private intervalDelay;
+
+  private pollTokens: Set<string>;
+
+  private fetchGasEstimates;
+
+  /**
+   * Creates a GasFeeController instance
+   *
+   */
+  constructor({
+    interval = 15000,
+    messenger,
+    state,
+    fetchGasEstimates = defaultFetchGasEstimates,
+  }: {
+    interval?: number;
+    messenger: RestrictedControllerMessenger<
+      typeof name,
+      GetGasFeeState,
+      GasFeeStateChange,
+      never,
+      never
+    >;
+    state?: Partial<GasFeeState>;
+    fetchGasEstimates?: typeof defaultFetchGasEstimates;
+  }) {
+    super({
+      name,
+      metadata,
+      messenger,
+      state: { ...defaultState, ...state },
+    });
+    this.intervalDelay = interval;
+    this.fetchGasEstimates = fetchGasEstimates;
+    this.pollTokens = new Set();
+  }
+
+  async getGasFeeEstimatesAndStartPolling(
+    pollToken: string,
+  ): Promise<GasFeeState | undefined> {
+    let gasEstimates;
+    if (this.pollTokens.size > 0) {
+      gasEstimates = this.state;
+    } else {
+      gasEstimates = await this._fetchGasFeeEstimateData();
+    }
+
+    this._startPolling(pollToken);
+
+    return gasEstimates;
+  }
```

Comment:
I would not expect start polling to return the `gasEstimates` since those will be updated in the state and we are following a reactive paradigm, the values should always be read from the state as it is the source of truth. 

I would expect though that start polling returns the `pollToken` or a fresh `pollToken` in case the argument is falsy, consider this code:

```js
// "Mounting / Unmounting / changing pollToken" effect
useEffect(() => {
  const token = getGasFeeEstimatesAndStartPolling(pollToken);
  setPollToken(token);
  return () => {
	if (token) {
      disconnectPoller(token);
    }
  }
}, [setPollToken, pollToken]);
```

this would work in the case `pollToken` is set or not:
```js
const [pollToken, setPollToken] = useState(null);
// or 
const [pollToken, setPollToken] = useState(txHash);
```

In order to do so:
- start polling with an existing poll token is a no op
- start polling with a falsy poll token returns a new poll token

I'd encourage the use of this pattern because we minimize UI <-> Controllers polling issues.